### PR TITLE
Feature/programming exercises/override results

### PR DESCRIPTION
**Pull request by krusche**
Fixes issue #1162 

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
Instructors should be able to override programming exercise results
I also refactored some duplicated code and improved the logic on the server so that only tutors can save / submit / override their own assessment before the assessment due date and so that instructors can always do it for modeling, file upload, text and programming exercises

- [x] There is still one TODO with respect to code duplication in my changes, that need to be fixed
- [x] Also, it would be great to explicitly test the new logic with server integration tests for all modeling, file upload, text and programming exercises

### Steps for Testing
1. Add manual result for a programming exercise
2. Instructors should always be able to override it
3. Tutors should not be able to override it
4. Also check this for modeling, file upload and text exercises

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 1]
```diff
@@ -94,6 +96,31 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
...
+    public boolean isAllowedToOverrideExistingResult(@NotNull Result existingResult, Exercise exercise, User user, boolean isAtLeastInstructor) {
+        // if the assessor is null, the user can save / submit / override the existing result
+        final var isAssessor = existingResult.getAssessor() == null || user.equals(existingResult.getAssessor());
+        if (existingResult.getCompletionDate() == null) {
+            // if the result exists, but was not yet submitted (i.e. completionDate not set), the tutor and the instructor can override, independent of the assessment due date
+            return isAssessor || isAtLeastInstructor;
+        }
+        // if the result was already submitted, the tutor can only override before a potentially existing assessment due date
+        var assessmentDueDate = exercise.getAssessmentDueDate();
+        final var isBeforeAssessmentDueDate = assessmentDueDate != null && ZonedDateTime.now().isBefore(assessmentDueDate);
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java]: What if there is no assessment due date? I noticed this when testing, for exercises without assessment due date, a tutor cannot override his own assessment. The override button can be clicked but the tutor just receives a `403` from the backend then.

File: src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java [hunk 2]
```diff
@@ -35,4 +60,67 @@ void validateExercise(Exercise exercise) throws BadRequestAlertException {
...
+     *
+     * If the result does not yet exist or is not yet submitted, this method returns true
+     *
+     * @param submission the submission that might include an existing result which would include information about the assessor
+     * @param exercise the exercise to which the submission and result belong and which potentially includes an assessment due date
+     * @param user the user who initiates a request
+     * @param isAtLeastInstructor whether the given user is an instructor for the given exercise
+     * @return true of the the given user can override a potentially existing result
+     */
+    protected boolean isAllowedToOverrideExistingResult(Submission submission, Exercise exercise, User user, boolean isAtLeastInstructor) {
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java]: Maybe the name of this method and the one below can be slightly changed to differentiate them from the method with the same name in the assessment service that has slightly different preconditions? For the latter, a concrete existing result is passed in as a parameter while the former do not actually require a result to exist to return a useful boolean value. Maybe something like `isAllowedToCreateOrUpdateResult` or `isAllowedToCreateOrOverrideResult`?

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java]: That's intended. The general idea is that as soon as a student can see the result, it should not change anymore. So if the assessment due date has passed or there is no assessment due date, tutors cannot change the assessment any more after submitting it.

If the buttons are still shown on the client side, we should hide them

File: src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java (hunk 2 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/AssessmentResource.java]: good point, will adapt it

**Comment by krusche** [comment]
> Will we add assessment due dates to programming exercises? Because with these changes, tutors are not able to add manual results at all since programming exercises do not have an assessment due date.

Thanks for raising this point. To make the exercise experience consistent, we should add the assessment due date in the future for manual / semi-automatic programming exercises as well.

For now - to stay consistent with the current behavior - we should probably leave the logic as is for programming exercises and allow the same tutor to edit a result even "after" the assessment due date.

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 1 above)
Comment (madwau) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java]: Ok, makes sense, thanks for clarifying.

**Comment by krusche** [comment]
@lovaiible thanks for testing, I fixed this and this should work again

**Comment by lovaiible** [comment]
@krusche Yes, the manual result works now and I can also override them as an instructor. Thanks for fixing the issue :) Overriding the other exercise type submissions also works fine.

When I click on `Update Result` in the tutor exercise dashboard and programming exercise scores page (e.g. https://artemistest.ase.in.tum.de/#/course/19/exercise/1462/dashboard), I receive the following console error:
image

**Pull request merged**